### DISCUSSION TITLE
No way to include custom <!--[if mso]> rules?

### DISCUSSION THREAD

**Discussion by dchesbro** [resend/react-email#1389] [Ideas]
### The problem

I'm trying to add Outlook specific markup to my email templates, and so far have been unable to find any way to render MSO rules like `<!--[if mso]><![endif]-->` in the final HTML output.

Including Outlook rules within brackets (i.e., `{'<!--[if mso]><![endif]-->'}`) just prints the text of the comment, and trying suggestions online like using `dangerouslySetInnerHTML` just results in an empty `<div></div>`.

Is there any currently accessible way to include rules like this in templates?

### Why not do this manually?

Results in comment being printed:

```
{'<!--[if mso]><![endif]-->'}
```

Results in empty `<div></div>`:

```
<div
  dangerouslySetInnerHTML={{
    __html: '<!--[if mso]><![endif]-->',
  }}
></div>
```

**Comment by ronakmehtav** [comment]
Do you still need mso tag's ? like for styling reason or the structure is not correct in outlook ?

**Comment by derozn** [comment]
+1 to this. The structure does not cater for outlook

**Comment by DamienDeloubes** [comment]
Adding a way to include comments in the final output would be ideal 🙌

**Comment by teouxcreative** [comment]
This seems like a must have to me.
I'm just getting started with creating email templates but what I've made with React Email completely breaks in Outlook which is disappointing.

It could be due to my lack of experience, and there may be ways to handle it I'm not aware of, but the only thing I can think of right now would be to manually add the code afterwards which doesn't seem like a great workflow.

**Comment by mattiaz9** [comment]
I managed to do it this way:
```tsx
import React, { PropsWithChildren } from "react"
import { renderToString } from "react-dom/server"

type TargetProps = PropsWithChildren<{
  logic: string
}>

export function Target({ logic, children }: TargetProps) {
  return (
    <span
      dangerouslySetInnerHTML={{
        __html: `
          <!--[if ${logic}]>
          ${renderToString(children as any)}
          <![endif]-->
        `,
      }}
    />
  )
}
```

then:
```tsx
<Target logic="mso">
  <Text>this is outlook</Text>
</Target>
```

**Comment by DamienDeloubes** [comment]
> I managed to do it this way:
> 
> ```tsx
> import React, { PropsWithChildren } from "react"
> import { renderToString } from "react-dom/server"
> 
> type TargetProps = PropsWithChildren<{
>   logic: string
> }>
> 
> export function Target({ logic, children }: TargetProps) {
>   return (
>     <span
>       dangerouslySetInnerHTML={{
>         __html: `
>           
>         `,
>       }}
>     />
>   )
> }
> ```
> 
> then:
> 
> ```tsx
> <Target logic="mso">
>   <Text>this is outlook</Text>
> </Target>
> ```

@mattiaz9 Can you provide it in a CodeSandbox or show the html output? I tried this but the comments are not included in the output HTML.

(Tried it with the render utility and the dev environment)

**Comment by scotthebertpt** [comment]
If it helps anyone else trying to do this, I managed to accomplish this by simply running an additional post-processing step on the string that react email's `render` method returns. You should be able to extend this for your use case as needed.

**My approach:**

I wrote a simple helper function that looks for data attributes and replaces with html comments in the final string output.

```ts
function renderComments(htmlString: string): string {
    const updatedHtmlString = htmlString
      .replace(
        /<div\s+data-comment-start="([^"]+)"\s*><\/div>/g,
        '<!--[if $1]><!-- -->',
      )
      .replace(/<div\s+data-comment-end="([^"]+)"\s*><\/div>/g, '<!--<![$1]-->')
      .replace(
        /<div\s+data-comment-mso-start="([^"]+)"\s*><\/div>/g,
        '<!--[if $1]>',
      )
      .replace(
        /<div\s+data-comment-mso-end="([^"]+)"\s*><\/div>/g,
        '<![$1]-->',
      );

    return updatedHtmlString;
  }
  ```
combine this with the render call and you're in business.

```ts
const initialMarkup = render(<Email />, { pretty: true });
const markup = renderComments(initialMarkup);
```

for additional convenience I added the following helper component

```tsx
import React from "react";

interface Props {
  children: React.ReactNode;
  comment: string;
  msoOnly?: boolean;
}

export default function HtmlConditionalComment({
  children,
  comment,
  msoOnly,
}: Props) {
  return msoOnly ? (
    <>
      <div data-comment-mso-start={comment} />
      {children}
      <div data-comment-mso-end="endif" />
    </>
  ) : (
    <>
      <div data-comment-start={comment} />
      {children}
      <div data-comment-end="endif" />
    </>
  );
}
```

Then in your email template just

```tsx
<HtmlConditionalComment comment="!mso">
  {/* non mso stuff */}
</HtmlConditionalComment>

<HtmlConditionalComment comment="mso" msoOnly={true}>
{/* mso stuff */}
</HtmlConditionalComment>
```

Hope that helps

**Reply by WillsB3** [reply]
Hey @scotthebertpt,

How did you integrate this solution with React.emails `email export` command? Did you make a custom version of that command and just use that or did you not need to use the `export` command for your setup?

**Reply by RexYao97** [reply]
this is a good way

**Comment by gabrielmfern** [comment]
Would be nice if we had another component to render specifically on outlook.

**Reply by ronakmehtav** [reply]
Well the new outlook, is much more similiar to it's browser version, and understand css better, and don't require MSO tags as well. So if you are targeting only the latest, then it doesn't need a separate component. Otherwise,  how/ where/ who would hold that complexity cause the styling would the be from the user so the api would be complex.

**Reply by WillsB3** [reply]
I think it's a bit too early to be targeting "latest outlook only"...

**Comment by mathsnunes** [comment]
hi guys,

Do you have any news on this, or do we think we should implement some of the suggestions above?

**Comment by gabrielmfern** [comment]
@mathsnunes 
> Do you have any news on this, or do we think we should implement some of the suggestions above?

Not yet anything new on this. Still something we have in mind though. I'd recommend you go with one of the solutions mentioned.

**Comment by ankitmishra88** [comment]
If it helps someone, here's what I did for our project, we created a components MSO and NOMSO
```
import * as React from 'react';
export const NOMSO: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return (
        <div data-no-mso="1">
            {children}
        </div>
    )
}

export const MSO: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return (
        <div data-mso="1">
            {children}
        </div>
    )
}
```

then replaced this  to be wrapped with MSO tags
```
export const replaceNoMsoTags = (html: string): string => {
    // Replace data-no-mso divs with MSO conditional comments
    const noMsoRegex = /<div\s+data-no-mso="1"[^>]*>([\s\S]*?)<\/div>/gi;
    return html.replace(noMsoRegex, (match, innerContent) => {
        return `<!--[if !mso]><!-->${innerContent}<!--<![endif]-->`;
    });
}

export const replaceMsoTags = (html: string): string => {
    // Replace data-no-mso divs with MSO conditional comments
    const noMsoRegex = /<div\s+data-mso="1"[^>]*>([\s\S]*?)<\/div>/gi;
    return html.replace(noMsoRegex, (match, innerContent) => {
        return `<!--[if !mso]><!-->${innerContent}<!--<![endif]-->`;
    });
}

export const getEmailContent = async (content: React.ReactElement) => {
    let html = await pretty(await render(content)); 
    html = replaceNoMsoTags(html);
    html = replaceMsoTags(html);
    
    return html;
}
```

**Comment by stewartjarod** [comment]
I ran into the same problem and noticed a few of us in this thread arrived at the same solution independently — data-attribute markers + regex post-processing on the render() output.

The core issue is that React can't emit HTML comments. But custom HTML elements (names with a hyphen like `<mso-expr>`) pass through every React renderer untouched. React treats them as web components and renders them literally. react-email's `render()` does zero HTML filtering, so they survive the full pipeline.

I packaged this up as [`react-email-mso`](https://www.npmjs.com/package/react-email-mso). The API is a single `<Outlook>` component:

```tsx
import { Outlook, processConditionals } from 'react-email-mso';
import { render } from '@react-email/render';

const Email = () => (
  <Outlook fallback={<table><tbody><tr><td width="600">Ghost table</td></tr></tbody></table>}>
    <div style={{ maxWidth: 600 }}>Modern layout</div>
  </Outlook>
);

const html = processConditionals(await render(<Email />));
```

Children are the modern default, `fallback` is what Outlook gets. Also supports standalone mode (`<Outlook>`, `<Outlook not>`) and version targeting (`<Outlook expr="gte mso 9">`).

Ships with `BulletproofButton` (VML for Outlook, CSS for modern) and `Columns` (ghost tables for Outlook, inline-block for modern) built on the same primitives.

~20 lines of core code, zero dependencies, works with react-email's existing `render()`.

npm: https://www.npmjs.com/package/react-email-mso
GitHub: https://github.com/stewartjarod/react-email-mso

**Reply by stewartjarod** [reply]
happy to make a PR into react-email to add this support directly.

**Reply by e1himself** [reply]
Thank you Jarod! 

Great job making it a reusable package! :clap: 

There's one more feature that can come really handy for emails markup that, as I believe, is not currently supported by your `react-email-mso` package: [MSO ghost tables](https://stackoverflow.design/email/base/mso/#ghost-tables)

This is how I've achieved it in my codebase. 

It's relying on my `RawHtml` helper component that works similar to your `Outlook` component, but only renders the actual raw HTML markup - having no assumptions of what it's for, leaving more room for manoeuvres: 

```tsx
export function MsoConditionalWrapper(props: {
    condition: string;
    wrapper: (props: { children: ReactNode }) => ReactNode;
    children: ReactNode;
}) {
    return (
        <>
            <RawHtml html={`<!--[if ${props.condition}]>`} />
            {props.wrapper({
                children: (
                    <>
                        <RawHtml html="<![endif]-->" />
                        {props.children}
                        <RawHtml html={`<!--[if ${props.condition}]>`} />
                    </>
                ),
            })}
            <RawHtml html="<![endif]-->" />
        </>
    );
}
```

You can then use it for generating ghost tables markup like this:

```tsx
<MsoConditionalWrapper
    condition="mso | IE"
    wrapper={(wrapped) => (
        <table role="presentation" cellspacing="0" cellpadding="0" border="0" width="100%">
           <tr>
               <td width="340">{wrapped.children}</td>
           </tr>
        </table>
    )}
>
  <div>Hello from within a ghost table</div>
</MsoConditionalWrapper>
```

**Reply by stewartjarod** [reply]
oh very neat. I will look into this a bit more. 
Ghost tables are definitely the end goal usually.

**Comment by stewartjarod** [comment]
@gabrielmfern would love your thoughts on this when you get a chance — happy to open a PR against canary if the approach looks good.

**Comment by stewartjarod** [comment]
@e1himself — your wrapper-pattern insight shipped in [0.7.0](https://www.npmjs.com/package/react-email-mso). Used it internally to simplify `<Columns>`:

```tsx
<GhostTable
  wrapper={(slot) => (
    <table role="presentation" cellSpacing={0} cellPadding={0} border={0} width="100%">
      <tbody><tr>{slot}</tr></tbody>
    </table>
  )}
>
  {columns.map(/* each column is itself a nested GhostTable wrapping a td */)}
</GhostTable>
```

The mechanism: render the wrapper JSX with a unique slot marker, run it through `renderToStaticMarkup`, split the HTML string on the marker, emit each half inside its own `<mso-expr>` conditional. Children then sit between the two MSO-only blocks, visible universally — same shape you showed with `MsoConditionalWrapper` + `RawHtml`.

The old `<Columns>` emitted spurious `<div>` wrappers inside the conditionals (Outlook forgave it, but not technically valid nested-in-table HTML). New output:

```html
<!--[if mso]><table ...><tbody><tr><td width="280" valign="top"><![endif]-->
<div style="display:inline-block;...">column content</div>
<!--[if mso]></td><![endif]-->
...
<!--[if mso]></tr></tbody></table><![endif]-->
```

Kept `GhostTable` internal-only for now. `Outlook fallback` covers the common case and I'd rather not grow the public API before there's clear user demand for the lower-level primitive. But the pattern is there if anyone wants to extend it.

Commit: https://github.com/stewartjarod/react-email-mso/commit/3464024
